test/spmd/tensor/test_pointwise_ops.py

```diff
@@ -8,94 +12,270 @@
 )
 from spmd.tensor.dispatch import OpSchema
 
+from spmd.tensor import distribute_tensor
 from spmd.tensor.ops.pointwise_ops import pointwise_rule
 from spmd import DeviceMesh, DTensor
-from spmd.tensor.placement_types import Shard, Replicate, _Partial, DTensorSpec
+from spmd.tensor.placement_types import (
+    Shard,
+    Replicate,
+    _Partial,
+    DTensorSpec,
+    Placement,
+)
 from torch.distributed.distributed_c10d import ReduceOp
 from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
 
+import torch.utils._pytree as pytree
+
+
+def deepcopy_convert_to_dtensor(
+    val: Any,
+    device_mesh: DeviceMesh,
+    placements: Sequence[Placement],
+) -> Any:
+    """
+    Recursively convert (over Sequence and Dict types) Tensors into DTensors.
+
+    :param device_mesh: the DeviceMesh to use.
+    :param placements: the Placement list to use.
+    :return: the transformed structure.
+    """
+
+    def f(x):
+        if isinstance(x, Tensor) and not isinstance(x, DTensor):
+            return distribute_tensor(
+                x,
+                device_mesh=device_mesh,
+                placements=placements,
+            )
+        return x
+
+    return pytree.tree_map(f, [val])[0]
+
+
+def deepcopy_convert_from_dtensor(val: Any) -> Any:
+    """
+    Recursive convert any DTensor to local Tensor.
+
+    :param val: the structure to coerce.
+    :return: the coerced structure.
+    """
+
+    def f(x):
+        if isinstance(x, DTensor):
+            return x.redistribute(
+                device_mesh=x.device_mesh,
+                placements=[Replicate()],
+            ).to_local()
+        return x
+
+    return pytree.tree_map(f, [val])[0]
+
 
 class DistElementwiseOpsTest(DistTensorTestBase):
+    def _compare_pairwise_ops(
+        self,
+        *,
+        device_mesh: DeviceMesh,
+        placements: Sequence[Placement],
+        op: Callable,
+        pre_op_reset: Optional[Callable] = None,
```

Comment:
this is a generalized refactoring of code that was already there. calling a reset callback is strictly more flexible than hardcoding some specialized behavior, and we should always select the composition answer over the specialized case answer.

Reply:
`pre_op_reset` seems to be not a generalized refactoring, it looks to me just a random seed "reset" name because there's no other reset can be done effectively, if you want to generalize it, maybe rename it to sth like `pre_op_fn`.

Reply:
done.